File: android/core/protocol/src/main/kotlin/app/hapi/protocol/wire/DecryptedMessage.kt
package app.hapi.protocol.wire

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * One stored chat message as returned by `GET /api/sessions/:id/messages` and
 * the `message-received` SSE event. Mirrors `DecryptedMessageSchema`
 * (`shared/src/schemas.ts`).
 *
 * [content] is the raw role-wrapped envelope — decoding it into renderable
 * blocks is the chat-pipeline milestone (`docs/api/client-contract/messages.md`),
 * not a wire concern.
 *
 * [invokedAt] is tri-state on purpose (`docs/api/client-contract/pagination.md`):
 * explicit `null` = the message is still queued; a number = when the agent
 * consumed it; **absent** = already invoked (pre-V8 hubs omit the field). A
 * plain `Long?` would silently merge "queued" with "legacy invoked", so this
 * class carries a hand-written serializer.
 */
@Serializable(with = DecryptedMessageSerializer::class)
data class DecryptedMessage(
    val id: String,
    val seq: Long? = null,
    val localId: String? = null,
    val content: JsonElement = JsonNull,
    val createdAt: Long,
    val invokedAt: OptionalField<Long?> = OptionalField.Absent,
    val scheduledAt: Long? = null,
    val deliveryState: String? = null,
) {
    /** `invokedAt` collapsed to a plain value (both absent and `null` → null). */
    val invokedAtOrNull: Long? get() = invokedAt.valueOrNull()

    /** Display position: `invokedAt ?? createdAt` (see pagination.md "Position key"). */
    val positionAt: Long get() = invokedAtOrNull ?: createdAt
}

internal object DecryptedMessageSerializer : KSerializer<DecryptedMessage> {
    override val descriptor: SerialDescriptor =
        SerialDescriptor("app.hapi.protocol.wire.DecryptedMessage", JsonObject.serializer().descriptor)

    override fun deserialize(decoder: Decoder): DecryptedMessage {
        val input = decoder as? JsonDecoder
            ?: throw SerializationException("DecryptedMessage supports JSON only")
        val obj = input.decodeJsonElement() as? JsonObject
            ?: throw SerializationException("DecryptedMessage must be a JSON object")
        return DecryptedMessage(
            id = obj["id"].stringOrNull
                ?: throw SerializationException("DecryptedMessage.id must be a string"),
            seq = obj["seq"].longOrNull,
            localId = obj["localId"].stringOrNull,
            content = obj["content"] ?: JsonNull,
            createdAt = obj["createdAt"].longOrNull
                ?: throw SerializationException("DecryptedMessage.createdAt must be a number"),
            invokedAt = if (obj.containsKey("invokedAt")) {
                OptionalField.Present(obj["invokedAt"].longOrNull)
            } else {
                OptionalField.Absent
            },
            scheduledAt = obj["scheduledAt"].longOrNull,
            deliveryState = obj["deliveryState"].stringOrNull,
        )
    }

    override fun serialize(encoder: Encoder, value: DecryptedMessage) {
        val output = encoder as? JsonEncoder
            ?: throw SerializationException("DecryptedMessage supports JSON only")
        output.encodeJsonElement(buildJsonObject {
            put("id", JsonPrimitive(value.id))
            put("seq", value.seq?.let(::JsonPrimitive) ?: JsonNull)
            put("localId", value.localId?.let(::JsonPrimitive) ?: JsonNull)
            put("content", value.content)
            put("createdAt", JsonPrimitive(value.createdAt))
            if (value.invokedAt is OptionalField.Present) {
                put("invokedAt", value.invokedAt.value?.let(::JsonPrimitive) ?: JsonNull)
            }
            value.scheduledAt?.let { put("scheduledAt", JsonPrimitive(it)) }
            value.deliveryState?.let { put("deliveryState", JsonPrimitive(it)) }
        })
    }
}

/**
 * User-message attachment descriptor (`AttachmentMetadataSchema`). Also the
 * shape sent back in `POST /sessions/:id/messages` bodies.
 */
@Serializable
data class AttachmentMetadata(
    val id: String,
    val filename: String,
    val mimeType: String,
    val size: Long,
    val path: String,
    val previewUrl: String? = null,
)

File: ios/Packages/HapiKit/Sources/HapiProtocol/Models/DecryptedMessage.swift
import Foundation

/// One chat message as returned by `GET /api/sessions/:id/messages` and the
/// `message-received` SSE event.
///
/// Mirrors `DecryptedMessageSchema` (`shared/src/schemas.ts`). `content` is
/// deliberately left as raw ``JSONValue``: it is the role-wrapped envelope
/// whose decode tree (`docs/api/client-contract/messages.md`) is ported in the
/// M2 chat milestones. Decoding a message is total — an absent or `null`
/// `content` degrades to `.null` instead of throwing.
public struct DecryptedMessage: Codable, Equatable, Sendable {
    /// Server uuid. Optimistic rows use the client `localId` until echoed.
    public var id: String
    /// Per-session insert counter; `null` on optimistic rows.
    public var seq: Int?
    /// Client-generated id for optimistic reconciliation.
    public var localId: String?
    /// Role-wrapped envelope, kept wire-verbatim (decoded by the chat pipeline).
    public var content: JSONValue
    /// Hub receive time (epoch ms).
    public var createdAt: Int
    /// When the agent consumed the message (epoch ms). `nil` covers both
    /// wire-`null` (still queued) and an absent key (pre-V8 hubs stamped
    /// rows as already invoked without the field). Current hubs always send
    /// the key, so treating `nil` as "queued" is correct against them; if
    /// pre-V8 hub support is ever needed, the M2 window store must
    /// distinguish explicit null (see pagination.md "Queued semantics").
    public var invokedAt: Int?
    /// Future-scheduled send time (epoch ms), when set.
    public var scheduledAt: Int?
    /// `indeterminate` means a steer outcome is unknown and needs explicit resolution.
    public var deliveryState: String?

    public init(
        id: String,
        seq: Int? = nil,
        localId: String? = nil,
        content: JSONValue = .null,
        createdAt: Int,
        invokedAt: Int? = nil,
        scheduledAt: Int? = nil,
        deliveryState: String? = nil
    ) {
        self.id = id
        self.seq = seq
        self.localId = localId
        self.content = content
        self.createdAt = createdAt
        self.invokedAt = invokedAt
        self.scheduledAt = scheduledAt
        self.deliveryState = deliveryState
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        seq = try container.decodeIfPresent(Int.self, forKey: .seq)
        localId = try container.decodeIfPresent(String.self, forKey: .localId)
        content = try container.decodeIfPresent(JSONValue.self, forKey: .content) ?? .null
        createdAt = try container.decode(Int.self, forKey: .createdAt)
        invokedAt = try container.decodeIfPresent(Int.self, forKey: .invokedAt)
        scheduledAt = try container.decodeIfPresent(Int.self, forKey: .scheduledAt)
        deliveryState = try container.decodeIfPresent(String.self, forKey: .deliveryState)
    }
}

/// Metadata describing an uploaded message attachment.
///
/// Mirrors `AttachmentMetadataSchema` (`shared/src/schemas.ts`). Used inside
/// user-message payloads and in the send-message request body.
public struct AttachmentMetadata: Codable, Equatable, Sendable {
    public var id: String
    public var filename: String
    public var mimeType: String
    /// Size in bytes.
    public var size: Int
    public var path: String
    /// Web-serving detail; present on wire payloads, absent from fixtures.
    public var previewUrl: String?

    public init(
        id: String,
        filename: String,
        mimeType: String,
        size: Int,
        path: String,
        previewUrl: String? = nil
    ) {
        self.id = id
        self.filename = filename
        self.mimeType = mimeType
        self.size = size
        self.path = path
        self.previewUrl = previewUrl
    }
}
